actions/src/openstack_check_actions.py

```diff
@@ -0,0 +1,277 @@
+                    rules["remote_ip_prefix"] == ip_prefix
+                    and rules["port_range_max"] == max_port
+                    and rules["port_range_min"] == min_port
+                ):
+                    return rules
+        return {}
+
+    @staticmethod
+    def _check_if_applied(rules: Dict, cloud, project):
+        rules_with_issues = []
+        with OpenstackConnection(cloud_name=cloud) as conn:
+            servers = conn.list_servers(
+                filters={"all_tenants": True, "project_id": project}
+            )
+
+        for server in servers:
+
+            with OpenstackConnection(cloud_name=cloud) as conn:
+                applied = conn.list_server_security_groups(server.id)
+
+            for serv_groups in applied:
+                if serv_groups["id"] == rules["security_group_id"]:
+                    print("Rule is applied")
+                    rules_with_issues.append(
+                        {
+                            "dataTitle": {"id": server["id"]},
+                            "dataBody": {
+                                "proj_id": project,
+                                "sec_id": serv_groups["id"],
+                                "applied": "Yes",
+                            },
+                        }
+                    )
+                else:
+                    rules_with_issues.append(
+                        {
+                            "dataTitle": {"id": server["id"]},
+                            "dataBody": {
+                                "proj_id": project,
+                                "sec_id": serv_groups["id"],
+                                "applied": "no",
+                            },
+                        }
+                    )
+                    print("Rule not applied to " + server["id"])
+        return rules_with_issues
+
+    def security_groups_check(
+        self,
+        cloud_account: str,
+        max_port: int,
+        min_port: int,
+        ip_prefix: str,
+        project_id=None,
+        all_projects=False,
+    ):
+        """
+        Starts the script to check projects, can choose to go through all projects or a single project.
+
+        Params:
+        Cloud (Required): The name of the cloud to open from clouds.yaml
+
+        Requires one of:
+            all_projects (optional): Boolean for checking all projects, defaults to False
+            project (optional): ID of project to check
+        """
+
+        # all_servers.filter(location.project.id=project)
+
+        # print(servers.get_all_servers())
+        rules_with_issues = {
+            "title": "Server {p[id]} has an incorrectly configured server group",
+            "body": "Project id: {p[proj_id]}\nSecurity Group ID: {p[sec_id]}\n Is applied: {p[applied]}",
+            "server_list": [],
+        }
+
+        if all_projects:
+            with OpenstackConnection(cloud_name=cloud_account) as conn:
+                projects = conn.list_projects()
```

Comment:
This can be simplified by treating a list as 1 or more objects (so a single project would be handled by the same code):
```python
if all_projects:
   with OpenstackConnection(...)
        projects = conn.list_projects()
else:
    projects = [project]

for project in projects:
  output = self._check_project_loadbalancers(
  project=project["id"],
  cloud=cloud_account,
  max_port=max_port,
  min_port=min_port,
  ip_prefix=ip_prefix,
                )
  rules_with_issues["server_list"].extend(output)